xbmc/platform/linux/input/LibInputHandler.cpp

```diff
@@ -0,0 +1,249 @@
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "LibInputHandler.h"
+#include "LibInputKeyboard.h"
+#include "LibInputPointer.h"
+#include "LibInputTouch.h"
+
+#include "utils/log.h"
+
+#include <algorithm>
+#include <fcntl.h>
+#include <linux/input.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+static int open_restricted(const char *path, int flags, void *user_data)
+{
+  (void)user_data;
+
+  int fd = open(path, flags);
+
+  if (fd < 0)
+  {
+    CLog::Log(LOGERROR, "%s - failed to open %s (%s)", __FUNCTION__, path, strerror(errno));
+    return -errno;
+  }
+
+  auto ret = ioctl(fd, EVIOCGRAB, (void*)1);
+  if (ret < 0)
+  {
+    CLog::Log(LOGDEBUG, "%s - grab requested, but failed for %s (%s)", __FUNCTION__, path, strerror(errno));
+  }
+
+  return fd;
+}
+
+static void close_restricted(int fd, void *user_data)
+{
+  (void) user_data;
+  close(fd);
+}
+
+static const struct libinput_interface m_interface =
+{
+  open_restricted,
+  close_restricted,
+};
+
+static void LogHandler(libinput *libinput, libinput_log_priority priority, const char *format, va_list args)
+{
+  (void)libinput;
+
+  if (priority == LIBINPUT_LOG_PRIORITY_DEBUG)
+  {
+    char buf[512];
+    int n = vsnprintf(buf, sizeof(buf), format, args);
+    if (n > 0)
+      CLog::Log(LOGDEBUG, "libinput: %s", buf);
+  }
+}
+
+CLibInputHandler::CLibInputHandler(CWinEventsLinux *winEvents)
+{
+  m_udev = udev_new();
+  if (!m_udev)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to get udev context for libinput", __FUNCTION__);
+    return;
+  }
+
+  m_li = libinput_udev_create_context(&m_interface, nullptr, m_udev);
+  if (!m_li)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to get libinput context", __FUNCTION__);
+    return;
+  }
+
+  libinput_log_set_handler(m_li, LogHandler);
+  libinput_log_set_priority(m_li, LIBINPUT_LOG_PRIORITY_DEBUG);
+
+  auto ret = libinput_udev_assign_seat(m_li, "seat0");
+  if (ret < 0)
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to assign seat", __FUNCTION__);
+
+  m_liFd = libinput_get_fd(m_li);
+
+  m_keyboard.reset(new CLibInputKeyboard(winEvents));
+  m_pointer.reset(new CLibInputPointer(winEvents));
+  m_touch.reset(new CLibInputTouch());
+}
+
+CLibInputHandler::~CLibInputHandler()
+{
+  if (m_li)
+    libinput_unref(m_li);
```

Comment:
`libinput_unref` accepts `NULL` as parameter, the check isn't needed. Same with `udev_unref`